File: app/src/main/java/br/com/edsilfer/moviedb/service/RestTemplate.java
package br.com.edsilfer.moviedb.service;


import android.content.Context;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONObject;

import br.com.edsilfer.moviedb.model.ResponseType;
import br.com.edsilfer.moviedb.model.ResponseWrapper;
import br.com.edsilfer.moviedb.model.enums.EventCatalog;

/**
 * Created by edgar on 17/02/2016.
 */
public class RestTemplate {

    public static final String TAG = "RestTemplate";

    private String mUrl;
    private JSONObject mPayload;
    private JsonObjectRequest mRequest;
    private RequestQueue mRequestQueue;

    public RestTemplate() {
        this.mUrl = "";
    }

    public void execute(Context context) {
        if (mRequestQueue == null) {
            mRequestQueue = Volley.newRequestQueue(context);
        }
        mRequestQueue.add(this.mRequest);
    }


    public void setBody(final EventCatalog event, int httpMethod) {
        this.mRequest = new JsonObjectRequest(
                httpMethod,
                this.mUrl,
                this.mPayload,
                new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        ResponseWrapper responseWrapper = new ResponseWrapper(response, ResponseType.SUCCESS);
                        NotificationCenter.INSTANCE.notify(event, responseWrapper);
                    }
                },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        ResponseWrapper response = new ResponseWrapper(error.getMessage(), ResponseType.ERROR);
                        NotificationCenter.INSTANCE.notify(event, response);
                    }
                });
    }

    public void setPayload(JSONObject mPayload) {
        this.mPayload = mPayload;
    }

    public void setUrl(String url) {
        this.mUrl = url;
    }

    public void cancelRequests() {
        if (mRequestQueue != null) {
            mRequestQueue.cancelAll(new RequestQueue.RequestFilter() {
                @Override
                public boolean apply(Request<?> request) {
                    return true;
                }
            });
        }
    }

}

File: app/src/main/java/br/com/edsilfer/moviedb/service/comm/RestTemplate.kt
package br.com.edsilfer.moviedb.service.comm

import android.content.Context

import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.VolleyError
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley

import org.json.JSONObject

import br.com.edsilfer.moviedb.model.enums.ResponseType
import br.com.edsilfer.moviedb.model.ResponseWrapper
import br.com.edsilfer.moviedb.model.enums.EventCatalog

/**
 * Handles common methods related to the assembly of a REST call
 */
class RestTemplate {

    private var mUrl: String? = null
    private var mPayload: JSONObject? = null
    private var mRequest: JsonObjectRequest? = null
    private var mRequestQueue: RequestQueue? = null

    // LIFECYCLE ===================================================================================
    init {
        this.mUrl = ""
    }

    // PUBLIC INTERFACE ============================================================================
    fun execute(context: Context) {
        if (mRequestQueue == null) {
            mRequestQueue = Volley.newRequestQueue(context)
        }
        mRequestQueue!!.add(this.mRequest!!)
    }

    fun setBody(event: EventCatalog, httpMethod: Int) {
        this.mRequest = JsonObjectRequest(
                httpMethod,
                mUrl,
                mPayload,
                Response.Listener<org.json.JSONObject> { response ->
                    val responseWrapper = ResponseWrapper(response, ResponseType.SUCCESS)
                    NotificationCenter.notify(event, responseWrapper)
                },
                Response.ErrorListener { error ->
                    val response = ResponseWrapper(error.message, ResponseType.ERROR)
                    NotificationCenter.notify(event, response)
                })
    }

    fun setPayload(mPayload: JSONObject) {
        this.mPayload = mPayload
    }

    fun setUrl(url: String) {
        this.mUrl = url
    }

    fun cancelRequests() {
        if (mRequestQueue != null) {
            mRequestQueue!!.cancelAll(RequestQueue.RequestFilter { true })
        }
    }
}
